### ISSUE TITLE
itk(GDCM|DCMTK)SeriesFileNames: memory leaks, bad behavior

### ISSUE THREAD

**Issue by PHLF** [InsightSoftwareConsortium/ITK#2735]
Hello,

   I may have not searched thoroughly but I am surprised not to have found an issue about this:

- `itkGDCMSeriesFileNames` seems to suffer from a huge memory leak: it relies on the `gdcm::SerieHelper` class which... has some issues like this [commented call to delete](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/ThirdParty/GDCM/src/gdcm/Source/MediaStorageAndFileFormat/gdcmSerieHelper.cxx#L112)
  And looking at GCDM API doc, its authors seem well aware of the situation:
  
![image](https://user-images.githubusercontent.com/10827301/133207975-a4a1bfde-5a24-4757-91ba-78619077d111.png)

But you know... [who read the doc anyway](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/IO/GDCM/include/itkGDCMSeriesFileNames.h#L206)? (Please don't take this too seriously ;) )

So out of luck and running out of time, instead of patching ITK's GDCM I tried to use `itkDCMTKSeriesFileNames` which does not suffer from memory leaks, but... when you are out of luck you should stop trying:

- `itkDCMTKSeriesFileNames` stores SeriesUID in a vector... which means that when you are querying Series UIDs you end up with duplicates (n files of the same series => n duplicated seriesUID): they should be stored in a set/unordered_set (or maybe in a hash map for linking series with their files). Not to mention that it lags behind `itkGDCMSeriesFileName` in terms of features:
  - The recursive flag is not used
  - The `AddSeriesRestriction` function is a NOOP
  - The series uids and filenames are recomputed at each call to `GetInputFileNames`/`GetFileNames("seriesUID")`/`GetSeriesUID` but are stored as member variables: they could be used as cache values

Thank you very much for your time and for your work on this library.

**Comment by dzenanz** [comment]
I don't think that memory leaks are known. @malaterre @issakomi correct me if I am wrong. @PHLF do you have a simple repro case for a memory leak?

**Comment by PHLF** [comment]
Something like:

```cpp
  using SeriesIdContainer = std::vector<std::string>;
  SeriesIdContainer seriesUIDs;

  
  auto nameGenerator = itk::DCMTKSeriesFileNames ::New();

  nameGenerator->SetUseSeriesDetails(true);
  nameGenerator->AddSeriesRestriction("0020|000d");
  nameGenerator->SetGlobalWarningDisplay(false);
  nameGenerator->SetRecursive(false);
  nameGenerator->SetLoadSequences(false);

  namespace fs = std::filesystem;

  // inDcmRootDir is a directory containing DICOM data subfolders
  for (auto const& dirEntry : fs::recursive_directory_iterator{inDcmRootDir})
  {
    if (!dirEntry.is_directory())
    {
      continue;
    }
    // I am pretty sure the leak happen during the following call as GDCMSeriesFileNames call "m_seriesHelper.Clear()"
    nameGenerator->SetInputDirectory(dirEntry.path());

    seriesUIDs = nameGenerator->GetSeriesUIDs();

    for(const auto& seriesUID : seriesUIDs)
    {
      nameGenerator->GetFileNames(seriesUID);
    }
  }
```

**Comment by PHLF** [comment]
But maybe I did not use the API properly and there is no leak at all.

**Comment by issakomi** [comment]
I have tried to run the above test with Valgrind on Debian 11, ITK-5.2.1 release. No leaks or other errors were detected. I don't the classes in my apps, maybe i am missing something. Could you reproduce leaks with Valgrind? And yes, probably it were not bad to review gdcmSeriesHelper.cxx.

**Comment by PHLF** [comment]
> Could you reproduce leaks with Valgrind?

When running my application with massif (valgrind heap analyzer) the allocated memory reach a limit value for an unknown reason. But when use the path of a folder containing several dicoms as a parameter to the `itk::GDCMSeriesFileNames::SetInputDirectory` I clearly see an amount of memory being allocated corresponding to the size of my folder's content. And this memory is never freed when I call `itk::GDCMSeriesFileNames::SetInputDirectory` with a different folder or simpler when I instantiate an  `itk::GDCMSeriesFileNames` owned by a smartpointer in a loop: destroying the itk::GDCMSeriesFileNames does not deallocate the memory corresponding to the DICOM contents.

Looking further in `gdcmSeriesHelper` used by `itk::GDCMSeriesFileNames`, on can clearly see that the WHOLE file content of each iterated `.dcm` file is copied into a FileList's node and these nodes are never destroyed as the call to the delete operator is commented in the `gdcmSeriesHelper::Clear` function.

**Comment by issakomi** [comment]
> I clearly see an amount of memory being allocated corresponding to the size of my folder's content. And this memory is never freed

OK, i have looked at it again, in fact, tried folder with DICOM subfolders, 1GB. The scan seems to stress the memory, it grows and grows, after it is done with one sub-folder, it start another and memory is not freed. Valgrind still doesn't detect a leak, because memory is freed at exit.

**Comment by hjmjohnson** [comment]
PR #6464 (issue #6463) fixes one of the parity gaps noted here: `itk::DCMTKSeriesFileNames` now orders slices geometrically (ImagePositionPatient projected on the slice normal), matching `GDCMSeriesFileNames`, instead of by InstanceNumber alone.

The four defects enumerated in this issue are **not** addressed by #6464 and remain open:

<details>
<summary>Still outstanding</summary>

1. `GetSeriesUIDs()` returns duplicates — `GetDicomData()` pushes the UID once per file rather than deduplicating into a set.
2. `m_Recursive` is stored but never used when scanning the directory.
3. `AddSeriesRestriction()` is effectively a NOOP.
4. Series UIDs / file names are recomputed on every `GetInputFileNames` / `GetFileNames` / `GetSeriesUIDs` call instead of being cached in the member variables that already exist.

**Correction re: the reported `gdcm::SerieHelper` leak — verified not a leak.** `FileList` is `std::vector< SmartPointer<FileWithName> >` (`gdcmSerieHelper.h:57`, since 2008). In `Clear()`, `l->clear()` releases every `SmartPointer` (freeing the `FileWithName` objects), `delete l` frees the container, and `~SerieHelper()` calls `Clear()` — so nothing leaks. The commented-out `//delete *it;` is dead code from the pre-`SmartPointer` era (it wouldn't compile against a `SmartPointer` value); misleading, but not a leak. So that item should be considered resolved/invalid.

</details>

These would make a good focused follow-up to bring `DCMTKSeriesFileNames` to full feature parity with `GDCMSeriesFileNames`.
